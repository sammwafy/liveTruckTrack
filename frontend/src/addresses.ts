import {configs} from './configs'
import fetch from 'cross-fetch'
export interface HereAddress {
  [addressPart: string]: string
}

export interface AddressSuggestion {
  locationId: string
  address: HereAddress
  countryCode: string
  label: string
}
export interface AddressInfo {
  location: Object
}

console.log(configs);

const addressOrdering = [
  'postalCode',
  'street',
  'houseNumber',
  'district',
  'city',
  'county',
  'state',
  'country'
]

const wholeEuropeISO3 =
  'AND,ALB,AUT,BIH,BEL,BGR,BLR,CHE,CYP,CZE,DEU,DNK,EST,ESP,FIN,FRO,FRA,GBR,GIB,GRC,HRV,HUN,IRL,IMN,ISL,ITA,LIE,LTU,LUX,LVA,MCO,MDA,MNE,MKD,MLT,NLD,NOR,POL,PRT,ROU,SRB,RUS,SWE,SVN,SVK,SMR,TUR,UKR,VAT,XKX'

export const buildAddressString = (
  address: HereAddress,
  depth = addressOrdering.length
): string => {
  const addressArr: string[] = []
  addressOrdering.forEach((element) => {
    if (
      addressArr.length < depth &&
      typeof address[element] === 'string' &&
      address[element] !== ''
    ) {
      addressArr.push(address[element])
    }
  })
  return addressArr.join(', ')
}

export const getSuggestions = async (
  query: string
): Promise<AddressSuggestion[]> => {
  try {
    const url = new URL(
      'https://autocomplete.geocoder.ls.hereapi.com/6.2/suggest.json'
    )
    const searchParams = new URLSearchParams({
      app_id: configs.HERE_APPID,
      apiKey: configs.HERE_APPCODE,
      query,
      country: wholeEuropeISO3,
      maxresults: '5',
      language: 'en'
    })
    url.search = searchParams.toString()
    const response: any = await fetch(url.toString())
    const { suggestions } = await response.json()

    if (!Array.isArray(suggestions)) {
      return []
    }
    return suggestions.map((suggestion: AddressSuggestion) => ({
      ...suggestion,
      label: buildAddressString(suggestion.address)
    }))
  } catch (error) {
    console.error(
      'An error occurred while trying to fetch address suggestions',
      error
    )
    return []
  }
}

export const getLocationInfo = async (labelQuery: string): Promise<AddressInfo[]> => {
  try {
    const url = new URL(
      'https://geocoder.ls.hereapi.com/6.2/geocode.json'
    )
    const searchParams = new URLSearchParams({
      searchtext: labelQuery,
      gen: '9',
      app_id: configs.HERE_APPID,
      apiKey: configs.HERE_APPCODE
    })
    url.search = searchParams.toString()

    const response = await fetch(url.toString(), {
      method: 'GET',
      mode: 'cors'
    })
    const locationData: any = await response.json()

    const location = locationData?.Response?.View[0]?.Result[0].Location?.NavigationPosition[0]

    return location
  } catch (error) {
    console.error(
      'An error occurred while trying to fetch address Info',
      error
    )
    return []
  }
}
