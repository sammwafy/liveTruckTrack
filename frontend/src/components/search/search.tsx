import { FC, useEffect, useState } from "react";
import { SearchWrapper } from "./styles/search.styled";
import { FaCheck, FaSearch } from "react-icons/fa";
import {
  AddressSuggestion,
  getLocationInfo,
  getSuggestions,
} from "../../addresses";
import NearestDrivers from "../drivers/nearestDrivers";

export interface SearchProps {
  mockSearchValue?: string;
}
export const Search: FC<SearchProps> = ({ mockSearchValue }) => {
  const [results, setResults] = useState<AddressSuggestion[]>([]);
  const [selectedLocation, setSelectedLocation] = useState("");
  const [locationInfo, setlocationInfo] = useState({});
  const [query, setQuery] = useState(mockSearchValue ?? "");
  const [suggestionView, setSuggestionView] = useState(false);

  useEffect(() => {
    let isMounted = true; // check if component is mounted -> fixes jest error ( should be fixed in react 18)
    getSuggestions(query)
      .then((data) => isMounted && setResults(data))
      .catch((error) => console.log(error));

    // clear up
    return () => {
      isMounted = false;
    };
  }, [query]);

  useEffect(() => {
    selectedLocation !== "" &&
      getLocationInfo(selectedLocation).then((result) =>
        setlocationInfo(result)
      );
  }, [selectedLocation]);

  const inputHandler = (selected: string): void => {
    setSelectedLocation(selected);
    setSuggestionView(false);
    setQuery(selected);
  };


  return (
    <div>
      <SearchWrapper data-testid="SearchWrapper">
        <div className="searchTitle">
          <h1>Search for an address</h1>
        </div>

        <div className="searchContainer">
          <div className="search">
            <i>{selectedLocation !== "" ? <FaCheck /> : <FaSearch />}</i>
            <input
              type="text"
              id="searchBar"
              placeholder="Pickup location..."
              onChange={(e) => {
                setQuery(e.target.value);
                setSuggestionView(true);
              }}
              value={query}
            />
          </div>
          {results.length > 0 && (
            <div className={`${suggestionView ? "searchResults" : "d-none"}`}>
              {Object.values(results).map((result, i) => {
                return (
                  <div
                    onClick={() => {
                      inputHandler(result.label);
                    }}
                    key={i}
                  >
                    {result.label}
                  </div>
                );
              })}
            </div>
          )}
        </div>
      </SearchWrapper>
      {Object.keys(locationInfo).length > 0 && (
        <NearestDrivers locationInfo={locationInfo} />
      )}
    </div>
  );
};
