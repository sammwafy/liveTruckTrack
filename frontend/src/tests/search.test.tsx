import { render } from '@testing-library/react'
import { getLocationInfo, getSuggestions } from '../addresses'
import { Search } from '../components/search/search'

it('should fetch and render', async () => {
  await new Promise(process.nextTick)
  const locationRes = await getLocationInfo('Berlin, Germany')
  const SuggestionsRes = await getSuggestions('Berlin, Germany')
  expect(locationRes).toMatchSnapshot()

  expect(SuggestionsRes).toMatchSnapshot()

  const { findByText, getByTestId } = render(
    <Search mockSearchValue='Berlin, Germany' />
  )
  expect(
    await findByText(
      '24848, Berlin, Potsdam, Klein Bennebek, Schleswig-Flensburg, Schleswig-Holstein, Germany'
    )
  ).toBeInTheDocument()

  // style component test
  const SearchWrapper = getByTestId('SearchWrapper')

  expect(SearchWrapper).toHaveStyle('padding: 16px 0 24px 0px;')
})
