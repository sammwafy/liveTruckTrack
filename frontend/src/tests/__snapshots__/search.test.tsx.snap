// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch and render 1`] = `
Object {
  "Latitude": 52.51605,
  "Longitude": 13.37691,
}
`;

exports[`should fetch and render 2`] = `
Array [
  Object {
    "address": Object {
      "city": "Berlin",
      "country": "Germany",
      "county": "Berlin",
      "postalCode": "10117",
      "state": "Berlin",
    },
    "countryCode": "DEU",
    "label": "10117, Berlin, Berlin, Berlin, Germany",
    "language": "en",
    "locationId": "NT_DKX2wfaWICS7JyfHmCOJhB",
    "matchLevel": "city",
  },
  Object {
    "address": Object {
      "country": "Germany",
      "county": "Berlin",
      "state": "Berlin",
    },
    "countryCode": "DEU",
    "label": "Berlin, Berlin, Germany",
    "language": "en",
    "locationId": "NT_MA6xiXs5hGsq-tRm65Wf-A",
    "matchLevel": "county",
  },
  Object {
    "address": Object {
      "country": "Germany",
      "state": "Berlin",
    },
    "countryCode": "DEU",
    "label": "Berlin, Germany",
    "language": "en",
    "locationId": "NT_yGnW7rZT3HnFM0EL.MXf8D",
    "matchLevel": "state",
  },
  Object {
    "address": Object {
      "city": "Seedorf",
      "country": "Germany",
      "county": "Segeberg",
      "district": "Berlin",
      "postalCode": "23823",
      "state": "Schleswig-Holstein",
    },
    "countryCode": "DEU",
    "label": "23823, Berlin, Seedorf, Segeberg, Schleswig-Holstein, Germany",
    "language": "en",
    "locationId": "NT_z98IMzwzx4B46-Frosm6PA",
    "matchLevel": "district",
  },
  Object {
    "address": Object {
      "city": "Klein Bennebek",
      "country": "Germany",
      "county": "Schleswig-Flensburg",
      "district": "Potsdam",
      "postalCode": "24848",
      "state": "Schleswig-Holstein",
      "street": "Berlin",
    },
    "countryCode": "DEU",
    "label": "24848, Berlin, Potsdam, Klein Bennebek, Schleswig-Flensburg, Schleswig-Holstein, Germany",
    "language": "en",
    "locationId": "NT_pFGfBfjhNCRlCWwNLLB3nD",
    "matchLevel": "street",
  },
]
`;
